File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res;
        int n = nums.size();
        res.push_back(1);
        for(int i=0;i<n-1;i++){
            res.push_back(res[i]*nums[i]);
        }
        int productSoFar = 1;
        for(int i=n-1;i>=0;i--){
            res[i] = res[i] * productSoFar;
            productSoFar *= nums[i];
        }

        return res;
    }
};

// 3 4 5 6

// 1 3 12 60
// 360 120 30 6 1

// 120 90 72 60

// 1 1
// 24 12 4 1

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        product_factor = 1
        res = nums.copy()

        for count in range(1,n):
            res[count] = res[count] * res[count-1]

        for count in range(n-1,0,-1):
            res[count] = res[count-1] * product_factor
            product_factor *= nums[count]

        res[0] = product_factor

        return res


            
        